libclang 6.0.1 (new formula)

albertosottile opened the pull request:
Provides a way to install just libclang without downloading and installing the whole llvm bottle.
It reduces build times and bottle sizes (270 MB vs 410 MB on Sierra).
Being keg-only it does not interfere with system or with full llvm installations.
It avoids potential library conflicts when only libclang.dylib and clang headers are required (see #29699).
Useful for building (and maybe using) qt and pyside without downloading llvm (see #29415).

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

albertosottile commented:
The test code is taken from [this gist](https://gist.github.com/quarnster/3686396). If possible, I would like @quarnster to authorize the use of his code in this repository and its potential release under the BSD-2-Clause license.  

Otherwise, we can stick to more traditional llvm-config based tests.

albertosottile commented:
P.S. I stripped the `head` section to pass `brew audit --new-formula`, but of course it is totally possible to put it back.

albertosottile commented:
Given that we did not get an answer from the gist author, I changed the test block with a new code heavily adapted by myself and inspired by a code released under the MIT license. This should eliminate any potential license issues.

albertosottile commented:
Some formulas that can be built and used with `libclang` instead of `llvm` (tested on 10.13.6):

```
castxml
creduce
nvc
pyside
rtags
woboc_codebrowser
```

DomT4 commented:
@BrewTestBot test this please

fxcoudert commented on Formula/libclang.rb:
```diff
@@ -0,0 +1,139 @@
+  fails_with :gcc
+  ("4.3".."4.6").each do |n|
+    fails_with :gcc => n
+  end
```
All those versions of GCC are very old and not carried by recent macOS and recent Homebrew anymore. Remove the whole block.

sjackman commented:
Is it possible for `llvm` to depend on `libclang`, so that both can be installed and linked?

DomT4 commented:
I'm leaning towards a 👍 here but I'll give it another day or two to let other maintainers discuss/comment as they wish to. The idea that 5/7 of the formulae that currently mandate a runtime `llvm` dependency could be switched over to use this much slimmer dependency is compelling.

fxcoudert commented on Formula/libclang.rb:
```diff
@@ -0,0 +1,139 @@
+      -DLLVM_INCLUDE_DOCS=OFF
+      -DLLVM_TARGETS_TO_BUILD=all
+    ]
+    args << "-DLIBOMP_ARCH=x86_64"
```
Inline all arguments below, no temporary `args` variable. (I am surprised `-DLIBOMP_ARCH=x86_64` is not auto-detected.)

DomT4 commented:
> Is it possible for llvm to depend on libclang, so that both can be installed and linked?

IIRC `llvm` contains `libclang` already, alongside a whole bunch of other stuff. I think OP is essentially saying that most of what other formulae need is in this formula, which could be used as a drop-in replacement where the full `llvm` package isn't necessary, given the full `llvm` package is a gigabyte heavier.

fxcoudert commented on Formula/libclang.rb:
```diff
@@ -0,0 +1,139 @@
+        clang_disposeIndex(index);
+        return 0;
+      }
+    EOS
```
Can we maybe trim down that example C source a bit?

albertosottile replied:
Done.

fxcoudert commented on Formula/libclang.rb:
```diff
@@ -0,0 +1,139 @@
+                           "libclangtest.cpp", "-o", "libclangtest"
+    testresult = shell_output("./libclangtest sample.cpp")
+
+    sorted_testresult = testresult.split("\n").sort.join("\n")
```
Why split, sort, and join if the result is a single line?

fxcoudert commented:
My only concern is that the name is misleading… we’re not only installing libclang, but also an actual clang binary, and more

DomT4 commented:
> My only concern is that the name is misleading… we’re not only installing libclang, but also an actual clang binary, and more

Aye, fair comment 👍. Open to suggestions if anyone has anything. `clang-lite` or `llvm-lite` or `llvm-clang` or something hopefully less slightly silly sounding than any of those.

MikeMcQuaid commented:
Personally I'm 👎 on this. It's annoying to have the extra disk space used but I don't think we should be optimising package management around disk space usage. Additionally, if you need to have something that depends on `llvm` and something that depends on `libclang` it will use _more_ disk space after this rather than less.

RandomDSdevel commented:
@MikeMcQuaid:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This new `libclang` formula (or whatever it ends up being named) might, in all probability, be needed in a PR I'm planning to submit that will restore the `qt` formula's `--with-docs` option (or resurrect something similar, such as a `--with-llvm`/`--with-libclang`/whatever option, in its place,) as @ilovezfs removed said pre-existing option in #29469 in response to my earlier report of it being broken in #29415.  Whether this PR gets merged or not, neither of us have yet managed to make Qt's documentation build properly with Homebrew's full LLVM keg.  There may still be a way to fix this (perhaps those two upstream patches I'm going to need anyway in addition to this PR might be enough to make `qdoc`'s build work well enough on their own,) but I haven't looked into doing that just yet.

albertosottile commented:
@fxcoudert 
I just pushed a commit according to your review, I hope I addressed all your concerns. I deleted 12 lines from the test code, I do not think it can be shrunk further without excessively compromising its legibility.

@MikeMcQuaid 
> Additionally, if you need to have something that depends on llvm and something that depends on libclang it will use more disk space after this rather than less.

To be honest, I do not know if any of the existing formulas actually need llvm. I am on vacation at the moment and I only have my MacBook Air with me, so I tested only the formulas that I posted before. It might be that other formulas can be built and used with libclang instead of llvm, and there is a strong chance that none of the current formulas actually need full llvm. In any case, reducing the number of formulas that use llvm reduces also the number of users that need to install it, also reducing the probability that someone needs to have both libclang and llvm in their disk. 

> It's annoying to have the extra disk space used but I don't think we should be optimising package management around disk space usage.

It's not just about disk space but also bandwidth and data usage, especially if llvm becomes a dependency of qt.

MikeMcQuaid commented:
> have yet managed to make Qt's documentation build properly with Homebrew's full LLVM keg

To me that points to removing the option rather than adding a new, duplicated formula.



> also reducing the probability that someone needs to have both libclang and llvm in their disk.

Currently that probability is 0% because there is no `libclang` formula 😉.

> It's not just about disk space but also bandwidth and data usage, especially if llvm becomes a dependency of qt.

The same argument applies. If you have a formula that does need to depend on LLVM and another that depends on this formula: you've just increased the disk, bandwidth and data usage for that user.

albertosottile commented:
> The same argument applies. If you have a formula that does need to depend on LLVM and another that depends on this formula: you've just increased the disk, bandwidth and data usage for that user.

That's true, but the vast majority of users will save disk space, bandwidth, and data usage. Also, does anyone know if there is any Formula that currently needs the full `llvm` keg?

albertosottile commented:
Update on other formulas compatibility with libclang: 
`zig` can be built with libclang and the included test passes. Now, of course, zig is a full language and it is not guaranteed that a single test covers all its requirements, but consider that zig itself provides pre-built "llvm+clang" library for Windows for their users [here](https://github.com/ziglang/zig/wiki/Building-Zig-on-Windows), in a 275 MB tar (_wink wink_). 

`emacs-clang-complete-async` is the last formula that depends on `llvm`, but unfortunately, it does not have a test block and I am not an emacs user so I cannot test it privately. Nevertheless, build runs just fine and [on their website](https://github.com/Golevka/emacs-clang-complete-async) only libclang is stated as a requirement. 

There are two other formulas with an optional `llvm` dependency: `doxygen` and `rust`, if you want I can also test those. In addition, there are other formulas that depend on older llvm versions, I am not sure what to do with those honestly. Besides optional dependencies and old versions, it seems that the proposed libclang formula supports all the existing formulas that currently depend on `llvm`.

About the name, I understand that libclang could appear confusing because a clang binary is included, even though I saw this name used by other parties for the same package (Qt in the first place). I like the proposed alternatives and I suggest also `clang-lib` for consideration.

albertosottile commented:
I apologize, I looked more carefully at sizes now that I am back at work and, apparently, I was comparing apples to oranges. While there is indeed a size reduction, is not as dramatic as I thought. 
Specifically, on Sierra (in mega binary bytes):

Formula | Bottled | Uncompressed
------------ | ------------- | -------------
`libclang` | 270 MB | 920 MB
`llvm` | 410 MB | 1375 MB

Looking only at those numbers, I have to agree with @MikeMcQuaid and concede that the size benefits alone do not justify the added hassle of maintaining two similar formulas. However, as mentioned before, `libclang` could still be useful when linking conflicts between system and `llvm` libraries arise (e.g. in `qt` and `pyside` formulas). 

In addition, I personally find quite puzzling that apparently none of the existing formulas need all the resources currently embedded in the `llvm` bottle. Maybe an alternative easier solution could be to make this extra stuff optional in the existing `llvm` formula (specifically, `clang-tools-extra, compiler-rt, libcxx, libunwind, lld, openmp, polly`). This would achieve both the results sought by this PR without adding a partially overlapping formula.

MikeMcQuaid commented:
Thanks for doing the research and being so open about this @albertosottile. I'm afraid as a result I definitely think it's worth closing this as-is.

> libclang could still be useful when linking conflicts between system and llvm libraries arise (e.g. in qt and pyside formulas).

Generally if this only applies to a few formulae we consider vendoring (e.g. using a `resource` block) in the relevant formula as a last resort.

> In addition, I personally find quite puzzling that apparently none of the existing formulas need all the resources currently embedded in the llvm bottle. Maybe an alternative easier solution could be to make this extra stuff optional in the existing llvm formula (specifically, clang-tools-extra, compiler-rt, libcxx, libunwind, lld, openmp, polly). This would achieve both the results sought by this PR without adding a partially overlapping formula.

I'm afraid we're moving in exactly the opposite direction here: 
https://github.com/Homebrew/homebrew-core/issues/31510. The desire is to have formulae with no options that provide all the necessary functionality over having multiple, overlapping formulae or formulae with options.

MikeMcQuaid commented:
I think the best bet here is for you to create your own tap in which you maintain a stripped down/option heavy LLVM. Sorry!

albertosottile commented:
To be honest, I do not need a stripped down LLVM, personally. With this PR, I was trying to solve [the issue](
https://github.com/Homebrew/homebrew-core/issues/29415#issuecomment-405893237) experienced by `qt --with-docs` and the linking conflict we found in `pyside`. The reduction in size was an added benefit that could have interested the community, or so I thought. Given your comments and #31510, I feel that you do not perceive that as a benefit at all. No hard feelings, though and, from my side, I apologize for not having checked sooner the uncompressed install size of this formula, but only its bottle size.

MikeMcQuaid commented:
There's absolutely no need to apologise @albertosottile; that you've done the research you have and it's all publicly documented is a genuinely useful contribution for future reference ❤️

DomT4 commented:
Thanks for the hard work here @albertosottile. I appreciate the time you put into this, despite the end result.
